Add support for additional modules in STM32U0.

ExaltZephyr opened the pull request:
This PR adds support for DMA, SPI, RNG, Crypto, USB, RTC, LPUART, LPTIM and WWDG modules in STM32U0.

ExaltZephyr committed:
dts: Add support for additional modules in STM32U0.

This commit Introduces DTS configurations for DMA,
SPI, RNG, Crypto, USB and RTC modules to enable
support in STM32U0.

Signed-off-by: Mohammad Badawi <zephyr@exalt.ps>
Signed-off-by: Sara Touqan <zephyr@exalt.ps>

ExaltZephyr committed:
boards: Enable additional modules for STM32 nucleo_u083rc.

This commit enables support for DMA,
SPI, RNG, Crypto, USB and RTC.

Signed-off-by: Mohammad Badawi <zephyr@exalt.ps>
Signed-off-by: Sara Touqan <zephyr@exalt.ps>

ExaltZephyr committed:
drivers: Add shared IRQ support for STM32U0 series in DMA driver.

This commit Modifies the DMA STM32 Kconfig to enable shared IRQ
support for the STM32U0 series.
This change ensures DMA channels with shared IRQs are properly
configured for stm32u0 devices.

Signed-off-by: Mohammad Badawi <zephyr@exalt.ps>
Signed-off-by: Sara Touqan <zephyr@exalt.ps>

djiatsaf-st commented on dts/arm/st/u0/stm32u0.dtsi:
```diff
@@ -273,6 +307,115 @@
+			clocks = <&rcc STM32_CLOCK(AHB1, 0U)>;
+			dma-channels = <7>;
+			dma-offset = <0>;
+			dma-requests= <12>;
```
When using stm32-dma-v2, I think it is not important to specify `dma-requests` since requests are handled in DMAMUX.

djiatsaf-st replied:
I took a look at your remark, and for me, there will be no issues because if we don't specify, the value 0 is given, which is not bad for this version of dma i think.

Besides, there is a confusion I found only for the `stm32-dma-v2` between `dma-requests` and `dma-channels`. It would have made more sense to me if it was only `dma-channels`. I think that work will be done later to avoid the confusion.

ExaltZephyr replied:
I see, but I remember that I had issues when I removed it through testing in past.
will try again to remove it and check what happens

ExaltZephyr replied:
When I removed dma-requests, I encountered issues.
I believe you intended me to remove dma-channels instead of dma-requests, similar to how it's handled on other boards.

I noticed that boards like stm32wb, stm32g0, stm32c0, and stm32wl do not mention dma-channels under the DMA nodes in their DTS configurations. 
To maintain consistency with these boards, I removed dma-channels and added dma-requests instead."

djiatsaf-st commented on dts/arm/st/u0/stm32u0.dtsi:
```diff
@@ -273,6 +307,115 @@
+			clocks = <&rcc STM32_CLOCK(AHB1, 1U)>;
+			dma-channels = <5>;
+			dma-offset = <7>;
+			dma-requests= <12>;
```
ditto

djiatsaf-st commented on dts/arm/st/u0/stm32u0.dtsi:
```diff
@@ -273,6 +307,115 @@
+			interrupts = <8 0>;
+			interrupt-names = "usb";
+			num-bidir-endpoints = <8>;
+			ram-size = <2048>;
```
According to the refman RM0503, Size of dedicated packet buffer memory SRAM for USB should be `1024` bytes instead of 2048

djiatsaf-st commented on boards/st/nucleo_u083rc/nucleo_u083rc.yaml:
```diff
@@ -14,5 +14,12 @@ supported:
   - i2c
   - pwm
   - usart
+  - spi
+  - dma
+  - usb
+  - rtc
+  - rng
+  - aes
+  - lptim
```
Nit  :  sort in alphabetical order with the rest

ExaltZephyr commented on boards/st/nucleo_u083rc/nucleo_u083rc.dts:
```diff
@@ -134,3 +151,48 @@
+};
+
+&spi1{
+	pinctrl-0 = <&spi1_sck_pa1 &spi1_miso_pa11 &spi1_mosi_pa12>;
```
I changed pins because I noticed that PA6 and PA7 conflict with the pins assigned to I2C2 in this file.

ExaltZephyr commented on boards/st/nucleo_u083rc/nucleo_u083rc.dts:
```diff
@@ -134,3 +151,48 @@
+zephyr_udc0: &usb {
+	clocks = <&rcc STM32_CLOCK(AHB1, 18U)>,
+		 <&rcc STM32_SRC_HSI48 CLK48_SEL(1)>;
+	pinctrl-0 = <&usb_dm_pa11 &usb_dp_pa12>;
```
yes correct, will change pins of spi.

djiatsaf-st commented on dts/arm/st/u0/stm32u0.dtsi:
```diff
@@ -273,6 +307,87 @@
+			#dma-cells = <3>;
+			reg = <0x40020800 0x400>;
+			interrupts = <11 0>;
+			dma-channels = <12>;
```
to be more compliant , STM32U073xx and STM32U083xx devices have 12 channels and STM32U031xx has 7 channels. 

We will have here  `dma-channels = <7>;`   and in stm32u073.dtsi  we will have at least 
```
dmamux1: dmamux@40020800 
{
	dma-channels = <12>;
	dma-requests = <76>;
};

```

ExaltZephyr commented:
Hi all,
the twister test "tests/drivers/adc/adc_api" keeps failing. 
During compilation, there's a warning in the adc_stm32 file.
I rebased on main, but the issue is still there. Is there an open bug for this?

djiatsaf-st commented:
> Hi all, the twister test "tests/drivers/adc/adc_api" keeps failing. During compilation, there's a warning in adc_stm32 file. I rebased on main, but the issue still there. Is there an open bug for this?

Hello @ExaltZephyr , 

Since `nucleo_u083rc` supports DMA (added in `nucleo_u083rc.yaml`), we need to add some configurations in `tests/drivers/adc/adc_api/boards/nucleo_u083rc.overlay` inside `adc1` node to solve the issue.

```
dmas = <&dmamux1 0 5 (STM32_DMA_PERIPH_RX | STM32_DMA_MEM_16BITS |
		STM32_DMA_PERIPH_16BITS)>;
	dma-names = "dmamux";
```

For test bench purposes on our side we use `stm32u083c_dk` instead of `nucleo_u083rc`, I would really appreciate it if you could add this support to `stm32u083c_dk`.

ExaltZephyr committed:
tests: drivers: adc: Enable DMA for nucleo_u083rc board for adc_api

Since DMA is supported by the nucleo_u083rc board, it should be enabled
in the board overlay for the ADC adc_api test, similar to the board
nucleo_h743zi.

Signed-off-by: Mohammad Badawi <zephyr@exalt.ps>
Signed-off-by: Sara Touqan <zephyr@exalt.ps>

erwango commented on boards/st/nucleo_u083rc/doc/index.rst:
```diff
@@ -176,10 +197,10 @@ Default Zephyr Peripheral Mapping:
+- SPI1_NSS : PA15
+- SPI1_SCK : PA1
+- SPI1_MISO : PA11
+- SPI1_MOSI : PA12
```
@ExaltZephyr Can you comment on this ?

ExaltZephyr replied:
@erwango 
I noticed that PA6 and PA7 conflict with the pins assigned to I2C2 in this file,so I made adjustments accordingly.
We now have two options: either retain these pin assignments and disable I2C2, or keep I2C2 enabled and reassign the pins ?

erwango replied:
If possible to have Arduino SPI, Arduino I2C and USB all together, please select this option.
Otherwise please comment I2C and we'll enable it in tests

ExaltZephyr replied:
I2C2 is not present on the Arduino connector, only I2C1 is available. 
Therefore, I will proceed with the second option and disable I2C2.

erwango commented on drivers/dma/Kconfig.stm32:
```diff
@@ -56,7 +56,7 @@ config DMA_STM32_SHARED_IRQS
 	depends on SOC_SERIES_STM32C0X || SOC_SERIES_STM32F0X || \
 		   SOC_SERIES_STM32G0X || SOC_SERIES_STM32L0X || \
-		   SOC_SERIES_STM32WB0X
+		   SOC_SERIES_STM32WB0X || SOC_SERIES_STM32U0X
```
Please keep alphabetical order

ExaltZephyr committed:
tests: drivers: i2c: Enable i2c2 for nucleo_u083rc board.

This commit enables I2C2 on nucleo_u083rc board
for i2c_target_api test.

Signed-off-by: Mohammad Badawi <zephyr@exalt.ps>
Signed-off-by: Sara Touqan <zephyr@exalt.ps>

erwango commented on boards/st/nucleo_u083rc/nucleo_u083rc.dts:
```diff
@@ -95,12 +112,14 @@
 	clock-frequency = <I2C_BITRATE_FAST>;
 };
 
+/*
 &i2c2 {
 	pinctrl-0 = <&i2c2_scl_pa7 &i2c2_sda_pa6>;
 	pinctrl-names = "default";
 	status = "okay";
 	clock-frequency = <I2C_BITRATE_FAST>;
 };
+*/
```
Not quite clean for the reader.
Let's remove purely the node, if there is no use today, or provide a statement on the reason why it is commented out.

erwango reviewed (changes requested):
One very last comment for me. Otherwise, this is fine thanks.

ExaltZephyr commented:
@erwango 
PR jobs are now passing! 
Is there anything else pending on our end?
Just to note, I'm not authorized to merge.

erwango commented:
> Is there anything else pending on our end?

No, just waiting more approvals

> Just to note, I'm not authorized to merge.

For sure, me neither. We need to wait to have 2 (green) approvals and then that a release engineer (the ones with merge rights) passes by.
